zim/tmp-kiwix/cleanup-old-files.py

```diff
@@ -19,76 +19,54 @@
 import os
 import argparse
 import datetime
+from pathlib import Path
 
-def list_files_to_delete(folder_path, days):
-    """
-    List files that are older than a specified number of days.
 
-    Args:
-        folder_path (str): The path to the target folder.
-        days (int): Delete files older than this many days.
-
-    Returns:
-        list: A list of file paths to be deleted.
-    """
-    current_time = datetime.datetime.now()
-    files_to_delete = []
-
-    def list_files(folder_path):
-        for foldername, subfolders, filenames in os.walk(folder_path):
-            for filename in filenames:
-                file_path = os.path.join(foldername, filename)
-                yield file_path
-
-    for file_path in list_files(folder_path):
-        # Get the last modification time of the file
-        file_time = datetime.datetime.fromtimestamp(os.path.getmtime(file_path))
-
-        # Calculate the difference in days
-        days_difference = (current_time - file_time).days
-
-        if days_difference > days:
+def list_files_to_delete(folder_path: Path, days: int) -> list[Path]:
+    """List files that are older than a specified number of days."""
+    minimum_time = (datetime.datetime.now() - datetime.timedelta(days=days)).timestamp()
+    files_to_delete: list[Path] = []
+    for file_path in folder_path.rglob("*"):
+        if file_path.stat().st_mtime < minimum_time:
             files_to_delete.append(file_path)
 
     return files_to_delete
 
-def list_directories_to_delete(folder_path, files_to_delete):
-    """
-    List empty subdirectories or subdirectories containing only files to be deleted.
 
-    Args:
-        folder_path (str): The path to the target folder.
-        files_to_delete (list): A list of file paths to be deleted.
+def list_directories_to_delete(
+    folder_path: Path, files_to_delete: list[Path]
+) -> list[Path]:
+    """List empty subdirectories or subdirectories containing only files to be deleted."""
+    empty_directories_to_delete: list[Path] = []
+
+    for file in files_to_delete:
+        parent = Path(file).parent
 
-    Returns:
-        list: A list of empty subdirectories to be deleted.
-    """
-    empty_directories_to_delete = []
+        # Check recursively if the parent directory is already empty or contains only
+        # files/directories that will be deleted
+        while all(
```

Comment:
It's probably OK but this is just too complicated for me to understand… which I think defeats the maintenance goal.

Can you rewrite it in a less-elegant but simpler, easier to read way?
Also, I think we should bound this to the passed folder (which can end up being empty). We don't want to delete it.

Here's a version I find easier to understand

```py
def list_directories_to_delete(
    folder_path: Path, files_to_delete: list[Path]
) -> list[Path]:
    """List empty subdirectories or subdirectories containing only files to be deleted."""
    folder_candidates: set[Path] = set()

    # loop over files (deepest first) to add folders (no i/o needed)
    for file in sorted(files_to_delete, key=lambda p: len(p.parts), reverse=True):
        for parent in file.parents:
            # make sure we only add child folders
            if parent.is_relative_to(folder_path) and parent != folder_path:
                folder_candidates.add(parent)

    # loop over unique list of folder (i/o to get list of files for each folder)
    for folder in list(folder_candidates):
        # check if there are file on disk in folder that are not to be deleted
        if [1 for file in folder.iterdir() if file not in files_to_delete]:
            folder_candidates.remove(folder)

    return list(folder_candidates)
```

Reply:
Too lazy to check 😉 Anyway you code is clear now